component/org.wso2.carbon.identity.oauth2.grant.token.exchange/src/main/java/org/wso2/carbon/identity/oauth2/grant/token/exchange/internal/TokenExchangeServiceComponent.java

```diff
@@ -37,4 +48,125 @@ protected void deactivate(ComponentContext ctxt) {
+            TokenExchangeComponentServiceHolder.getInstance().setUserOperationEventListeners(
+                    new TreeMap<Integer, UserOperationEventListener>());
+        }
+
+        TokenExchangeComponentServiceHolder.getInstance().putUserOperationEventListener(
+                userOperationEventListenerService.getExecutionOrderId(), userOperationEventListenerService);
+    }
+
+    protected void unsetUserOperationEventListenerService(
+            UserOperationEventListener userOperationEventListenerService) {
+        if (userOperationEventListenerService != null &&
+                TokenExchangeComponentServiceHolder.getInstance().getUserOperationEventListeners() != null) {
+            TokenExchangeComponentServiceHolder.getInstance().removeUserOperationEventListener(
+                    userOperationEventListenerService.getExecutionOrderId());
+            TokenExchangeComponentServiceHolder.getInstance().setUserOperationEventListenerCollection(null);
+        }
+    }
+
+    @Reference(
+            name = "identity.user.profile.mgt.component",
+            service = FederatedAssociationManager.class,
+            cardinality = ReferenceCardinality.MANDATORY,
+            policy = ReferencePolicy.DYNAMIC,
+            unbind = "unsetFederatedAssociationManagerService"
+    )
+    protected void setFederatedAssociationManagerService(FederatedAssociationManager
+                                                                 federatedAssociationManagerService) {
+
+        TokenExchangeComponentServiceHolder.getInstance().setFederatedAssociationManager(
+                federatedAssociationManagerService);
+    }
+
+    protected void unsetFederatedAssociationManagerService(FederatedAssociationManager
+                                                                   federatedAssociationManagerService) {
+
+        if (log.isDebugEnabled()) {
+            log.debug("Federated Association Manager Service is unset in the Application Authentication Framework " +
+                    "bundle");
+        }
+        TokenExchangeComponentServiceHolder.getInstance().setFederatedAssociationManager(null);
+    }
+
+    public static Collection<UserOperationEventListener> getUserOperationEventListeners() {
+        Map<Integer, UserOperationEventListener> userOperationEventListeners =
+                TokenExchangeComponentServiceHolder.getInstance().getUserOperationEventListeners();
+        Collection<UserOperationEventListener> userOperationEventListenerCollection =
+                TokenExchangeComponentServiceHolder.getInstance().getUserOperationEventListenerCollection();
+        if (userOperationEventListeners == null) {
+            userOperationEventListeners = new TreeMap<>();
+            TokenExchangeComponentServiceHolder.getInstance().setUserOperationEventListeners(
+                    userOperationEventListeners);
+        }
```

Comment:
I was having some issues with some listeners not getting set. Solution was to set userOperationEventListenerCollection to null whenever a listner service was set and repopulate the collection when getting. This approach was used in [user-account-association](https://github.com/wso2-extensions/identity-user-account-association) component so i used the same

Reply:
This needs to be fixed. however due to the urgency to merge the pr we will proceed for now since this is not a blocking issue. I will send a followup PR to fix this issue.